ContainerTop: harden

kolyshkin opened the pull request:
In case a user specifies arguments such as -C or f, optimization
introduced by commit a076bad is not used, and so the code falls back
to list all the processes and filter those with PIDs of the
container specified.

In case a user also specifies -o with some fields that may contain
spaces, and those fields are before the PID field (such as -o comm,pid),
PID filtering is broken. In the worst case scenario, command can
contain numeric arguments which are then parsed as PIDs, leading to
showing some processes not belonging to the container specified.

To fix, we need to make sure that PID field is either the first one
or at least before any other fields that may contain spaces. Here
we assume this is the case for any ps built-in formats; we only
need to fix it for the case when a custom format is specified.

So, we need to check if ps arguments contains custom formatting
options (-o/o/--format). If true:

1. Prepend `-opid` to ps arguments (so the first field is PID).
2. Always use the first field to obtain PID for filtering purposes.
3. Remove the first field from the output.

This fix looks easy in theory; in practice, though, due to variety
of ways ps can parse its arguments. Because of that, function
`isCustomFormat()` is not that trivial, plus a unit test case
is added to validate it works as expected.
    
This fixes two bugs:
 - wrongly listing processes not belonging to a container;
 - inability to use -o if it does not contain pid.

In addition, now, since we are adding our own PID column, any
tricks that a malicious user might do with column naming (such
as adding `-ouid=PID`) won't work anymore, so we can drop
the option validation (introduced in PR #24358) entirely, as well
as its test case.

Big thanks to https://github.com/lifubang for analysis, ideas and
patches (see https://github.com/moby/moby/pull/37748).

Test cases are available in the second commit.

kolyshkin committed:
ContainerTop: harden

In case a user specifies arguments such as -C or f, optimization
introduced by commit a076bad is not used, and so the code falls back
to list all the processes and filter those with PIDs of the
container specified.

In case a user also specifies -o with some fields that may contain
spaces, and those fields are before the PID field (such as -o comm,pid),
PID filtering is broken. In the worst case scenario, command can
contain numeric arguments which are then parsed as PIDs, leading to
showing some processes not belonging to the container specified.

To fix, we need to make sure that PID field is either the first one
or at least before any other fields that may contain spaces. Here
we assume this is the case for any ps built-in formats; we only
need to fix it for the case when a custom format is specified.

So, we need to check if ps arguments contains custom formatting
options (-o/o/--format). If true:

1. Prepend `-opid` to ps arguments (so the first field is PID).
2. Always use the first field to obtain PID for filtering purposes.
3. Remove the first field from the output.

This fix looks easy in theory; in practice, though, due to variety
of ways ps can parse its arguments. Because of that, function
`isCustomFormat()` is not that trivial, plus a unit test case
is added to validate it works as expected.

This fixes two bugs:
 - wrongly listing processes not belonging to a container;
 - inability to use -o if it does not contain pid.

In addition, now, since we are adding our own PID column, any
tricks that a malicious user might do with column naming (such
as adding `-ouid=PID`) won't work anymore, so we can drop
the option validation entirely, as well as its test case.

Big thanks to https://github.com/lifubang for analysis and
preliminary patches (see https://github.com/moby/moby/pull/37748).

Integration test comes in the next commit.

[v2: don't add pid if it's the first field already,
     add pid field even for the first ps (with q) run]
[v3: add o and --format handling, remove validatePSArgs]
[v4: make sure PID is the first column if we added it]
[v5: improve error return]
[v6: check for "o" in the middle of the argument]
[v7: better check for "o", added a unit test]
[v8: rewrite the changelog message]
[v9: fix the option parser]
[v10: only check for customFormat if custom args are given]

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin commented:
Wait a minute, I need to cover the `docker top $ID -ocmd` scenario (when `q id,id,...` is working).

kolyshkin commented:
@lifubang @thaJeztah PTAL

TODO: integration tests (I can't think of making a good unit test for this).

kolyshkin committed:
integration/container: Add ContainerTop tests

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

lifubang commented:
@kolyshkin , busybox support multi -o.
And I fix some more errors.
Please check it, thanks.

lifubang commented:
I'm so sorry that I don't test carefully. This patch can't fix the issue fully. I will think more.
@kolyshkin , you need to think more too.
We just filter -o o --format, but there are other situations:
ps eocmd

lifubang commented on daemon/top_unix.go:
```diff
@@ -163,11 +155,27 @@ func (daemon *Daemon) ContainerTop(name string, psArgs string) (*container.Conta
+	addPID := false
+	// Check whether o/-o/--format option is specified by a user
+	for _, arg := range args {
+		if strings.HasPrefix(arg, "-o") || strings.HasPrefix(arg, "o") || strings.HasPrefix(arg, "--format") {
```
docker top redisr eopid,cmd,%cpu
This situation can't work.

kolyshkin commented:
> ps eocmd

Oh yes, right, thanks for noticing this @lifubang! This looks more and more like a can of worms -- once you open it, you can't close it back :(

kolyshkin commented:
> We just filter -o o --format, but there are other situations:
> ps eocmd

@lifubang Looks like it's pretty simple: we need to always prepend `-o pid` to the list of arguments, always use it, and always remove it! Let me update the patch

kolyshkin commented:
> we need to always prepend -o pid to the list of arguments

I am very stupid :( of course we can't do that if there is no list specified, otherwise the output might only contain PID.

So, we go back to the previous solution, but also look `o<field>` even it is in the middle. This is complicated :(

kolyshkin commented:
Say `ps -A ocmd` does contain custom fields, but `ps -C ocmd` does not!

lifubang commented on daemon/top_unix.go:
```diff
@@ -130,6 +126,53 @@ func psPidsArg(pids []uint32) string {
+			return false
+		}
+
+		switch opt {
```
No filter for eocmd,ppid,pid
```
root@dockerdemo:~/moby# ps -C redis-server eocmd,ppid,pid
CMD                          PPID   PID
redis-server *:28808        19958 20005
redis-server *:6379 *:6379  22824 22839
redis-server *:6379         23391 23424
redis-server *:23391        23487 23521
root@dockerdemo:~/moby#
```

lifubang commented:
May be my original solution?

kolyshkin commented:
OK I think I nailed it this time.
The core of it is `isCustomFormat()` which tries to figure out whether some form of `-o` was given. I have added a unit test for that, please see `TestContainerTopIsCustomFormat()`.

To run the unit test:
```
cd daemon
go test -v -run CustomFormat .
```

@lifubang  Many thanks to you for coming up with ideas of what might break it -- can you please take another look at the code, try to break it maybe? If you can think of any other ways to cheat this, please add those to the unit test.

cpuguy83 commented:
Would we have way to correlate container PID with host PID?
If so we could just run `ps` in the container PID namespace and fix the pids.

kolyshkin commented:
> If so we could just run ps in the container PID namespace and fix the pids.

Yeah, I was thinking about that, too. In general, though, I think this is a very bad API (that allows a user to pass some arbitrary flags to some arbitrary binary, `ps`, that happens to be present on the host, and tries to filter out the results). This is the biggest problem with this API, but not the only one.

I think the real solution here is to deprecate this bad API, replacing it with something that collects information about container's processes, and sends it back in some more decent manner than a bunch of strings. The only problem is, I'm not sure how widespread is `ContainerTop()` use.

Maybe we can discuss this during the next maint meeting? // CC @thaJeztah

cpuguy83 commented:
Agreed, it's terrible.

kolyshkin committed:
ContainerTop: disallow --no-headers

[v2: add --no-header, found by accident]
[v3: add --no-heading, found in ps man page]
[v4: add other 5 forms, found in procps sources]

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

lifubang commented on daemon/top_unix.go:
```diff
@@ -130,6 +126,87 @@ func psPidsArg(pids []uint32) string {
+		if strings.HasPrefix(arg, "-o") || strings.HasPrefix(arg, "o") || strings.HasPrefix(arg, "--format") {
+			return true
+		}
+		if !strings.HasPrefix(arg, "--") && strings.Contains(arg, "o") {
```
It's too strict. such as:
ps -Cmongod

kolyshkin replied:
Fixed; my bad.

kolyshkin commented:
@lifubang as I said earlier (https://github.com/moby/moby/pull/37827#issuecomment-421170160), the long term course of action seems to be to deprecate this API, as it's practically impossible to get it 100% right. For the short term, I guess, we need to harden it as much as possible, I think we can do this (assuming `ps` won't accept any more flags that it currently does).

kolyshkin commented:
> I write a small lexer for analysis ps args options to resolve this problem.
> If you agree with me, you may cherry-pick.

It looks like you were relying on `ps --help all ` output or `man ps`. I took a look at ps source code, there are more options that it advertises (see the 3rd commit in this PR for example).

lifubang commented:
I quite agree with you.

lifubang commented:
Or you may call ps' interface of C language in golang?
https://salsa.debian.org/debian/procps/tree/master/ps

kolyshkin commented:
> Or you may call ps' interface of C language in golang?
you mean, for the purpose of figuring out whether the custom format is specified? It might work, but looks like an unnecessary complication.

One decent way to solve it would be to fork ps to make sure -q always works and is enforced.

kolyshkin commented:
> One decent way to solve it would be to fork ps to make sure -q always works and is enforced.

or leaving `-q` aside, add a similar option (or an env var) to enforce limiting ps output by a set of PIDs specified.

kolyshkin commented:
> fork ps ... to enforce limiting ps output by a set of PIDs specified.

This way, we will
1. Leave the API intact (although I'm sure this API is bad and need to be deprecated)
2. Allow all possible `ps` options to be used.
3. Remove a dependency on a third-party untrusted tool, `ps`, which we call and parse its output.

Now, our `ps` may even come as a go library (rather than a binary).

The downsides are:
1. Amount of work that need to be done is unknown (it looks simple, but who knows).
2. One more piece of customized software to maintain.

And it won't solve the other problem of the existing API, this one:
```
$ ID=$(docker run -d alpine sleep 100000)
$ docker top $ID -ocmd,pid
CMD                 PID
sleep               100000 31222
```

It can probably be solved by not splitting the output into lines (i.e. passing all the line in field[0]).

kolyshkin commented:
One alternative solution: don't split by whilespace, use positions. In other words, return to the previous code (banning `-o.*=PID` and `--no-headers`), look for positions of `PID` in the first column, use those to look for PIDs. If we'll make sure our own `ps` is used (say not `busybox ps`), then it can be done, as in its output the last digit of pid is always aligned with the last character from the `PID` header.

Implementation will look ugly though. I am more or less OK with that I have in this PR.

kolyshkin commented:
@AkihiroSuda PTAL (I see you worked on that before, https://github.com/moby/moby/pull/24358)

kolyshkin commented:
(rebased; changelog of the first commit updated)

kolyshkin commented:
CI failure in experimental is a network glitch.

CI failures in powerpc and z are the usual flaky suspects.

thaJeztah commented:
> Yeah, I was thinking about that, too. In general, though, I think this is a very bad API (that allows a user to pass some arbitrary flags to some arbitrary binary, ps, that happens to be present on the host, and tries to filter out the results). This is the biggest problem with this API, but not the only one.

Yes, definitely agree. The `top` implementation (think the first implementation was https://github.com/moby/moby/pull/1055/, followed by https://github.com/moby/moby/pull/1238 to allow passing arguments) was a looooong time ago, and just a quick 'n dirty way to get the basic functionality, and the easy way was to shell out and parse the output. It was never really "designed" for compatibility accross platforms, and no guarantee whatsoever w.r.t. which options would be supported (only a way to pass-through options to the `ps` command).

My worry with this PR is that we'll be adding _a lot of complexity_ to address for that original "quick 'n dirty" choice.

There have been several discussions around this to replace the ad-hoc / quick 'n dirty implementation with something that doesn't depend on an external binary, and where options are well-defined (I found that there actually was a proposal once to formalize options at the API level; https://github.com/moby/moby/pull/9232). I don't think anyone ever got around to working on that though.

Possible options;

- make sure the _default_ output works for all known situations
- explicitly call out that passing options _can_  break the output ("no guarantees whatsoever")
- decide _if_ we want an alternative implementation (weighing against how important the custom options are)

kolyshkin commented:
> My worry with this PR is that we'll be adding a lot of complexity to address for that original "quick 'n dirty" choice.

@thaJeztah I concur, this is a band-aid and unfortunately it is complicated.

Note the aim of this band-aid is to prevent "foreign" (i.e. ones not belonging to the container) processes showing in the output (as demonstrated by @lifubang). The issue discovered boils down to this: if you have a container and an access to `ContainerTop` (`docker top`) for the container, you have a way to obtain information about **any** host process in ContainerTop() output.

While we can start working on designing and implementing a better API to replace this one, we also need to assess how bad is the impact described above, and whether we consider it a security issue. If we do, something like this band-aid is still required until we remove this API.

kolyshkin commented:
First commit updated to not run the check if custom options are not provided (as per @thaJeztah suggestion). Also rebased

kolyshkin commented:
most elegant solution:
https://github.com/moby/moby/blob/2a63a5f7a5d7942d9848ca006c4a9945637ed0f3/daemon/top_windows.go#L27-L31

kolyshkin commented:
I think we can build something similar to https://github.com/moby/moby/pull/9232 (which is a great proposal!); the only thing that I would change is instead of specifying _fields_ to show, we specify the _groups_ of fields. To illustrate, here are a few groups and their meaning for Linux:
* **BASE**: all the *ids, state, cmdline;
* **CRED**: credentials (uid, gid, supplementary groups, capabilities);
* **STAT**: memory, cpu, io stats (aka "taskstats", see https://www.kernel.org/doc/Documentation/accounting/taskstats-struct.txt);
* ...

The idea to request groups (rather than specific fields) comes from experimental [task-diag](https://criu.org/Task-diag) Linux kernel API (a proposed alternative to /proc/PID).

Different platforms (Linux, Windows, ...) can fill this in in slightly different ways, with some fields being common between platforms, while some other being platform-specific.

`docker top` CLI can use the new interface.

Once having the new interface in place, we can deprecate this one.

I am also thinking about whether we should
a. have a separate new entry point for the new interface (like `/containers/ID/ps`), or
b. re-use the existing one (`/containers/ID/top`), or
c. add as an option to another existing API, say `containers/ID/json`.

I feel like (a).

lifubang commented:
@kolyshkin Please see https://github.com/moby/moby/pull/37748/commits/c49b4d2f493e0102488693d0e72d4ba9069d33d8
There is a new way for the short term. Please consider it.

thaJeztah commented on daemon/top_unix.go:
```diff
@@ -130,20 +126,145 @@ func psPidsArg(pids []uint32) string {
+// was given to ps
+func customFields(args []string) bool {
+	/*
+	 * ps allows for a few fancy ways to provide an argument:
```
nit: this comment could be part of the function's description instead of being in the function body

kolyshkin replied:
I am confident this is the best place for it. The function description tells *what* the function does, while the comments inside the function body tells *how* exactly it does that. This particular comment, I believe, is crucial to understanding how the function code works, but a user of the function doesn't have to know such gory details :)

Of course, if a function user is a fellow developer, the boundary is not so strict... but anyway

thaJeztah reviewed (approved):
LGTM

lifubang commented:
@thaJeztah, @kolyshkin  I think we still need more tests
Just now, I find a invalid test case for this patch.
```
root@dockerdemo:~/moby# docker top redisstor axZ
Error response from daemon: Unexpected pid '(enforce)': strconv.Atoi: parsing "(enforce)": invalid syntax
```
Because:
```
LABEL                             PID TTY      STAT   TIME COMMAND
docker-default (enforce)        26080 ?        Ss     0:00 /bin/bash /sto 6379
docker-default (enforce)        26123 ?        Sl     0:00 redis-server *:6379
```

As I mentioned in https://github.com/moby/moby/pull/37748#issuecomment-424361224, we can't list all of the test cases for this patch. So we need to change a way.
I think my last commit in https://github.com/moby/moby/pull/37748 may fix this issue with some small changes. Though it looks ugly, but it may safe.
I think try to parse ps options will make us involved too deep. If we can find a simple way to fix this issue,  we can spend more time to write the new api.
But I still need all of you review, and need more invalid test cases if exists.

lifubang commented:
@thaJeztah, @kolyshkin  I think we still need more tests
Just now, I find a invalid test case for this patch.
```
root@dockerdemo:~/moby# docker top redisstor -euroot
error during connect: Get http://%2Fvar%2Frun%2Fdocker.sock/v1.35/containers/redisstor/top?ps_args=-euroot: EOF
```

As I mentioned in https://github.com/moby/moby/pull/37748#issuecomment-424361224, we can't list all of the test cases for this patch. So we need to change a way.
I think my last commit in https://github.com/moby/moby/pull/37748 may fix this issue with some small changes. Though it looks ugly, but it may safe.
I think try to parse ps options will make us involved too deep. If we can find a simple way to fix this issue,  we can spend more time to write the new api.
But I still need all of you review, and need more invalid test cases if exists.

By the way, My commit also make output look more natually without any other changes:
```
root@dockerdemo:~/moby# docker top redisstor -euroot -ocmd,ppid,pid
CMD                   PPID                PID
/bin/bash /sto 6379   6844                6862
redis-server *:6379   6862                6910
```

kolyshkin commented:
> I think try to parse ps options will make us involved too deep.

I agree. ps option parsing is a nightmare -- for example, while `ps -eur` and `ps -euk` look similar, the first one treat `r` as an option, and the second one treat `k` as user name. It's a mess.

I will take a look at #37748; it looks like it is the way to go (for now) but needs some cleanups.

olljanat commented:
Derek add label: rebuild/*

cpuguy83 commented:
I feel like instead of using `ps` and trying to filter things it would be best to get the list of pids and whatever extra metadata we can provide (or users can ask for through options) and then let the client format it.

As you mentioned there is another PR that you think may be a better approach anyway, let's close this.
